src/linear_algebra/basis.lean

```diff
@@ -1424,3 +1424,51 @@ let ⟨q, hq⟩ := p.exists_is_compl in nonempty.intro $
   (prod_equiv_of_is_compl q p hq.symm)
 
 end division_ring
+
+section restrict_scalars
+
+variables [comm_ring R] [ring K] [nontrivial K] [add_comm_group V]
+variables [algebra R K] [module K V] [module R V]
+variables [is_scalar_tower R K V] [no_zero_smul_divisors R K] (b : basis ι K V)
+variable (R)
+
+open submodule
+
+/-- Let `b` be a `K`-basis of `V`. Let `R` be a comm_ring such that `algebra R K` with no zero
+smul divisors, then the submodule of `V` spanned by `b` over `R` admits `b` as a `R`-basis. -/
+noncomputable def basis.restrict_scalars : basis ι R (span R (set.range b)) :=
+basis.span (b.linear_independent.restrict_scalars (smul_left_injective R one_ne_zero))
+
+@[simp]
+lemma basis.restrict_scalars_apply (i : ι) : (b.restrict_scalars R i : V) = b i :=
+by simp only [basis.restrict_scalars, basis.span_apply]
+
+@[simp]
+lemma basis.restrict_scalars_repr_apply (m : span R (set.range b)) (i : ι) :
+  algebra_map R K ((b.restrict_scalars R).repr m i) = b.repr m i :=
+begin
+  suffices : finsupp.map_range.linear_map (algebra.linear_map R K) ∘ₗ
+      (b.restrict_scalars R).repr.to_linear_map
+      = ((b.repr : V →ₗ[K] (ι →₀ K)).restrict_scalars R).dom_restrict _,
+  { exact finsupp.congr_fun (linear_map.congr_fun this m) i, },
+  refine basis.ext (b.restrict_scalars R) (λ _, _),
+  simp only [linear_map.coe_comp, linear_equiv.coe_to_linear_map, function.comp_app, map_one,
+    basis.repr_self, finsupp.map_range.linear_map_apply, finsupp.map_range_single,
+    algebra.linear_map_apply, linear_map.dom_restrict_apply, linear_equiv.coe_coe,
+    basis.restrict_scalars_apply, linear_map.coe_restrict_scalars_eq_coe],
+end
+
+lemma basis.restrict_scalars_mem_span_iff (m : V) :
+  m ∈ span R (set.range b) ↔ ∀ i, b.repr m i ∈ set.range (algebra_map R K) :=
```

Comment:
This lemma name is a bit strange, as the statement doesn't use `restrict_scalars` at all.

Reply:
I agree. What about  `basis.mem_span_iff_repr_mem`?

Reply:
Sure; make sure to add a docstring explaining why the LHS is different to `basis.mem_span` (the basis is over `K` but the span is over `R`)

Reply:
I changed the name and added a docstring.